controller/controller_server/utils.py

```diff
@@ -8,43 +10,164 @@
 import controller.controller_server.messages as messages
 from controller.array_action.config import FC_CONNECTIVITY_TYPE, ISCSI_CONNECTIVITY_TYPE
 from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.controller_types import ArrayConnectionInfo, ObjectIdInfo, ObjectParameters
 from controller.controller_server.errors import ObjectIdError, ValidationException
 from controller.csi_general import csi_pb2
 
 logger = get_stdout_logger()
 
 
-def get_array_connection_info_from_secret(secrets):
+def _parse_raw_input(raw_input):
+    try:
+        parsed_input = json.loads(raw_input)
+    except json.decoder.JSONDecodeError:
+        raise ValidationException(messages.invalid_non_json_parameter_message.format(raw_input))
+    return parsed_input
+
+
+def _is_topology_match(system_topologies, node_topologies):
+    for topologies in system_topologies:
+        logger.debug(
+            "Comparing topologies: system topologies: {},"
+            " node topologies: {}".format(topologies, node_topologies))
+        if topologies.items() <= node_topologies.items():
+            return True
+    return False
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_system_info_by_topologies(secrets_config, node_topologies):
```

Comment:
`by` sounds like a mapping to me, consider using `for` instead

Reply:
I think it's correct, there is a map for a system by topologies and that what it getting

Reply:
does it return a dictionary where the topologies are the keys?

Reply:
its not returning map, but the system info that returns are mapped by topologies. (and AFAIK map and dictionary are equivalent.)

Reply:
not sure I follow, could you write an example of the structure here?

Reply:
the structure is the secret config.json

Reply:
I thought that in the config.json the system info is "mapped by" system id, not topologies

Reply:
ok, done. it's `for` now